middleware/mqtt_handler.py

```
@@ -296,8 +352,22 @@ def on_message(client: mqtt.Client, userdata: object, msg: mqtt.MQTTMessage) ->
         # Shared scanners (afc_stage/toolhead_stage) have no target to lock
         target = _get_scanner_target(scanner_cfg)
         if target and app_state.lane_locks.get(target):
-            logger.info(f"Ignoring scan on {target} (locked)")
-            return
+            if _should_auto_release_lock(target, payload):
+                incoming_uid = (payload.get("uid") or "").lower()
+                active_uid = app_state.active_spool_uids.get(target, "")
+                logger.info(
+                    f"Lock auto-release on {target}: idle printer, swap "
+                    f"{active_uid or '?'} → {incoming_uid}"
```

Ruff (0.15.12)

[warning] Logging statement uses f-string

(G004)